### PR TITLE
Truncate fixes

### PR DESCRIPTION
**Pull request by thomaso-mirodin**
Fix the issues in #3 + add some more testing

### CODE REVIEW

**Comment by thomaso-mirodin** [comment]
Okay @aybabtme, let me know if everything is still working on your end :D

**Comment by thomaso-mirodin** [comment]
@aybabtme Ping :D

File: tailf.go [hunk 1]
```diff
@@ -288,10 +298,93 @@ func (f *follower) updateFile() error {
...
+// it has regrown to be >= the same size as previously, the truncate
+// will be missed. tl;dr, don't use copy-truncate...
+func (f *follower) checkForTruncate() error {
+	f.mu.Lock()
+
+	fi, err := os.Stat(f.filename)
+
+	f.mu.Unlock()
+	if os.IsNotExist(err) {
+		return ErrFileRemoved{fmt.Errorf("File was removed: %v", f.filename)}
```
Comment (aybabtme) [inline tailf.go]: can you change the message to not be capitalized?

File: tailf.go [hunk 2]
```diff
@@ -288,10 +298,93 @@ func (f *follower) updateFile() error {
...
+	}
+	if err != nil {
+		return err
+	}
+
+	newSize := fi.Size()
+	if newSize < f.size {
+		err = ErrFileTruncated{fmt.Errorf("File (%s) was truncated", f.filename)}
```
Comment (aybabtme) [inline tailf.go]: same here

File: tailf.go [hunk 3]
```diff
@@ -201,36 +210,35 @@ func (f *follower) followFile() {
...
-		err = f.updateFile()
-	}
+	case isOp(ev, fsnotify.Chmod):
+		// Modified time on the file changed, noop
+		return nil
 
-	if err != nil {
-		log.Printf("handleEvent error: %v", err)
+	default:
+		panic(fmt.Sprintf("unknown event: %#v", ev))
```
Comment (aybabtme) [inline tailf.go]: this is scary, i'd prefer logging.

File: tailf.go [hunk 4]
```diff
@@ -288,10 +298,93 @@ func (f *follower) updateFile() error {
...
+	}
+	if err != nil {
+		return err
+	}
+
+	newSize := fi.Size()
+	if newSize < f.size {
+		err = ErrFileTruncated{fmt.Errorf("File (%s) was truncated", f.filename)}
+	} else {
+		err = nil
```
Comment (aybabtme) [inline tailf.go]: `err` can only be nil here, so this is not useful

File: tailf.go [hunk 5]
```diff
@@ -288,10 +298,93 @@ func (f *follower) updateFile() error {
...
+	previous_file, err := f.file.Stat()
+	if err != nil {
+		f.errc <- err
+	}
+
+	if err := f.watch.Add(f.filename); err != nil {
+		f.errc <- err
+	}
+
+	for current_file, err := os.Stat(f.filename); ; current_file, err = os.Stat(f.filename) {
```
Comment (aybabtme) [inline tailf.go]: you could make this the first call of the loop:

```go
for {
    currentFile, err := os.Stat(f.filename)
}
```

also, no `snake_case` in Go =P

File: tailf.go [hunk 6]
```diff
@@ -288,10 +298,93 @@ func (f *follower) updateFile() error {
...
+		err = nil
+	}
+
+	f.size = newSize
+	return err
+}
+
+// This is here for situations where the directory the watched file sits in can't be inotified on
+func (f *follower) pollForChanges() {
+	previous_file, err := f.file.Stat()
```
Comment (aybabtme) [inline tailf.go]: `snake_case` -> `camelCase`

File: tailf_test.go [hunk 7]
```diff
@@ -28,52 +38,69 @@ func canFollowFile(t *testing.T, filename string, file *os.File) error {
 
 	follow, err := tailf.Follow(filename, true)
 	if err != nil {
-		return fmt.Errorf("creating follower: %v", err)
+		t.Fatalf("Failed creating tailf.follower: '%v'", err)
```
Comment (aybabtme) [inline tailf_test.go]: invoking `t.Fatalf` won't run the `defer`s, so the temp file created for test will be left on disk.

also use uncapitalized messages to fit with the rest of the tests.

File: tailf_test.go [hunk 8]
```diff
@@ -28,52 +38,69 @@ func canFollowFile(t *testing.T, filename string, file *os.File) error {
...
 	}
 
 	go func() {
 		for _, str := range toWrite {
-			log.Printf("writing")
+			t.Logf("Writing %d bytes", len(str))
 			_, err := file.WriteString(str)
 			if err != nil {
-				t.Fatalf("failed to write to test file: %v", err)
+				t.Fatalf("Failed to write to file: '%v'", err)
```
Comment (aybabtme) [inline tailf_test.go]: same here, original formatting was fine, but `t.Fatal` was wrong.

File: tailf_test.go [hunk 9]
```diff
@@ -195,88 +215,275 @@ func canFollowFileFromEnd(t *testing.T, filename string, file *os.File) error {
...
 }
 
-func TestFollowTruncation(t *testing.T) { withTempFile(t, canFollowTruncation) }
+func TestFollowTruncation(t *testing.T) { withTempFile(t, time.Millisecond*150, canFollowTruncation) }
 
 func canFollowTruncation(t *testing.T, filename string, file *os.File) error {
 	follow, err := tailf.Follow(filename, false)
 	if err != nil {
-		t.Fatalf("failed to create follower: %v", err)
+		t.Fatalf("Failed creating tailf.follower: %v", err)
```
Comment (aybabtme) [inline tailf_test.go]: not a big fan of capitalization of all the messages, it's not a style that I favor and it's making the PR harder to read.

File: tailf.go (hunk 3 above)
Comment (thomaso-mirodin) [inline tailf.go]: Yeah, i dunno why I used panic, fixing.

File: tailf_test.go (hunk 7 above)
Comment (thomaso-mirodin) [inline tailf_test.go]: Good to know! I'll swap them back to returns

**Comment by thomaso-mirodin** [comment]
Fixed

**Comment by aybabtme** [comment]
thanks @thomaso-mirodin !

**Pull request merged**